libs/sdk-ui-ext/src/internal/dashboardEmbedding/KpiAlerts/DashboardItemWithKpiAlert.tsx

```diff
@@ -0,0 +1,288 @@
+import { IFilter } from "@gooddata/sdk-model";
+import { GoodDataSdkError, isNoDataSdkError } from "@gooddata/sdk-ui";
+import { Bubble, BubbleHoverTrigger } from "@gooddata/sdk-ui-kit";
+import cx from "classnames";
+import React, { Component, MouseEvent } from "react";
+import { FormattedMessage, WrappedComponentProps } from "react-intl";
+
+import { DashboardItemKpi } from "../DashboardItem/DashboardItemKpi";
+import { IKpiResult, IKpiAlertResult, KpiAlertOperationStatus } from "../types";
+import { isAlertingTemporarilyDisabledForGivenFilter } from "./utils/filterUtils";
+
+// adapted from jQuery:
+// https://github.com/jquery/jquery/blob/a503c691dc06c59acdafef6e54eca2613c6e4032/src/offset.js#L83-L97
+function getNodeDocumentRelativeOffsetTop(node: HTMLDivElement): number {
+    // Return zeros for disconnected and hidden (display: none) elements (gh-2310)
+    // Support: IE <=11+
+    // Running getBoundingClientRect on a
+    // disconnected node in IE throws an error
+    if (!node.getClientRects().length) {
+        return 0;
+    }
+
+    // Get document-relative position by adding viewport scroll to viewport-relative gBCR
+    const rect = node.getBoundingClientRect();
+    const win = node.ownerDocument.defaultView;
+    return rect.top + win.pageYOffset;
+}
+
+export interface IDashboardItemWithKpiAlertProps {
+    // KPI
+    kpi: IWidgetDefinition;
+    isLoading: boolean;
+    filters?: IFilter[];
+    kpiResult: IKpiResult | undefined;
+
+    // Alert
+    alert?: IWidgetAlertDefinition;
+    kpiAlertResult?: IKpiAlertResult | undefined;
+    alertExecutionError?: GoodDataSdkError;
+    isAlertExecutionLoading?: boolean;
+
+    canSetAlert?: boolean;
+    userWorkspaceSettings?: IUserWorkspaceSettings;
+    isAlertDialogOpen?: boolean;
+    isAlertHighlighted?: boolean;
+    isAlertLoading?: boolean;
+    alertSavingStatus?: KpiAlertOperationStatus;
+    alertUpdatingStatus?: KpiAlertOperationStatus;
+    alertDeletingStatus?: KpiAlertOperationStatus;
+    isAlertBroken?: boolean;
+
+    isReadOnlyMode?: boolean;
+
+    // Callbacks
+    onAlertDialogOpenClick: () => void;
+
+    contentClassName?: string;
+    kpiClassName?: string;
+    /**
+     * When true, alert will not be highlighted when triggered.
+     */
+    suppressAlertTriggered?: boolean;
+
+    children?: (params: { clientWidth: number }) => React.ReactNode;
+    renderHeadline?: () => React.ReactNode;
+    renderAlertDialog?: () => React.ReactNode;
+}
+
+interface IDashboardItemWithKpiAlertState {
+    isKpiAlertAfterSaving: boolean;
+    isKpiAlertAfterDeleting: boolean;
+    isAlertHighlighted: boolean;
+}
+
+export class DashboardItemWithKpiAlert extends Component<
+    IDashboardItemWithKpiAlertProps & WrappedComponentProps,
+    IDashboardItemWithKpiAlertState
+> {
+    static defaultProps: Partial<IDashboardItemWithKpiAlertProps & WrappedComponentProps> = {
+        isAlertHighlighted: false,
+        filters: [],
+        alertDeletingStatus: "idle",
```

Comment:
Maybe we should unify this with useCancelablePromise (where it's "pending")? "idle" is probably more appropriate.

Reply:
They are a bit different, I would prefer not mixing them.